### ISSUE TITLE
Multiple external ips are ignored

### ISSUE THREAD

**Issue by Northcode** [k3s-io/k3s#6032]
**Environmental Info:**
K3s Version: 
v1.24.3+k3s1

Node(s) CPU architecture, OS, and Version: 
```
5.10.0-17-amd64 #1 SMP Debian 5.10.136-1 (2022-08-13) x86_64 GNU/Linux
5.10.0-17-amd64 #1 SMP Debian 5.10.136-1 (2022-08-13) x86_64 GNU/Linux
```

Cluster Configuration:
1 master `node A`
1 agent `node B`
different networks connected via manual wireguard, flannel with vxlan over the wireguard channel

**Describe the bug:**
Specifying multiple external ips in `--node-external-ip` only uses the first one.
This is needed because I would like LoadBalancer services to have both the external interface ip for accessing it from the cluster externally, as well as the wireguard-ip to be able to access it from inside the vpn network.

If there is any way to work around this for LoadBalancer services I'd accept it as a workaround.

**Steps To Reproduce:**
- Installed K3s
- Set `--node-external-ip` on node A to `'1.2.3.4,10.0.0.1'` (where 1.2.3.4 would be the external interface, and 10.0.0.1 would be the vpn interface)
- Set `--node-external-ip` on node B to `'1.2.3.5,10.0.0.2'` (where 1.2.3.5 would be the external interface, and 10.0.0.2 would be the vpn interface)
- Install a LoadBalancer service like nginx-ingress

**Expected behavior:**
LoadBalancer service should have external-ips: `1.2.3.4,10.0.0.1,1.2.3.5,10.0.0.2`, being reachable from all external ips in the cluster

**Actual behavior:**
LoadBalancer service only has external-ips: `10.0.0.1,10.0.0.2`

**Additional context / logs:**

**Comment by manuelbuil** [comment]
> **Environmental Info:** K3s Version: v1.24.3+k3s1
> 
> Node(s) CPU architecture, OS, and Version:
> 
> ```
> 5.10.0-17-amd64 #1 SMP Debian 5.10.136-1 (2022-08-13) x86_64 GNU/Linux
> 5.10.0-17-amd64 #1 SMP Debian 5.10.136-1 (2022-08-13) x86_64 GNU/Linux
> ```
> 
> Cluster Configuration: 1 master `node A` 1 agent `node B` different networks connected via manual wireguard, flannel with vxlan over the wireguard channel
> 
> **Describe the bug:** Specifying multiple external ips in `--node-external-ip` only uses the first one. This is needed because I would like LoadBalancer services to have both the external interface ip for accessing it from the cluster externally, as well as the wireguard-ip to be able to access it from inside the vpn network.
> 
> If there is any way to work around this for LoadBalancer services I'd accept it as a workaround.
> 
> **Steps To Reproduce:**
> 
> * Installed K3s
> * Set `--node-external-ip` on node A to `'1.2.3.4,10.0.0.1'` (where 1.2.3.4 would be the external interface, and 10.0.0.1 would be the vpn interface)
> * Set `--node-external-ip` on node B to `'1.2.3.5,10.0.0.2'` (where 1.2.3.5 would be the external interface, and 10.0.0.2 would be the vpn interface)
> * Install a LoadBalancer service like nginx-ingress
> 
> **Expected behavior:** LoadBalancer service should have external-ips: `1.2.3.4,10.0.0.1,1.2.3.5,10.0.0.2`, being reachable from all external ips in the cluster
> 
> **Actual behavior:** LoadBalancer service only has external-ips: `10.0.0.1,10.0.0.2`
> 
> **Additional context / logs:**

Can you show me the annotations of one of your nodes?

**Comment by Northcode** [comment]
Here are the ones for the vps, real external ip replaced with 1.2.3.4:
```
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      alpha.kubernetes.io/provided-node-ip: 10.0.0.1
      csi.volume.kubernetes.io/nodeid: '{"zfs.csi.openebs.io":"northcode.no"}'
      flannel.alpha.coreos.com/backend-data: '{"VNI":1,"VtepMAC":"16:24:14:42:8d:a8"}'
      flannel.alpha.coreos.com/backend-type: vxlan
      flannel.alpha.coreos.com/kube-subnet-manager: "true"
      flannel.alpha.coreos.com/public-ip: 10.0.0.1
      k3s.io/external-ip: 1.2.3.4,10.0.0.1
      k3s.io/hostname: northcode.no
      k3s.io/internal-ip: 10.0.0.1
      k3s.io/node-args: '["server","--node-ip","10.0.0.1","--advertise-address","10.0.0.1","--node-external-ip","1.2.3.4,10.0.0.1","--flannel-iface","wg0","--write-kubeconfig-mode","664","--default-local-storage-path","/kube","--node-label","nettype=external","--disable","traefik"]'
      k3s.io/node-config-hash: SLN6277C5Y5LJOCLC2IQUIOYBU3FIRUNPY2PYXE2CZQMVFWMF5QA====
      k3s.io/node-env: '{"K3S_DATA_DIR":"/var/lib/rancher/k3s/data/1d787a9b6122e3e3b24afe621daa97f895d85f2cb9cc66860ea5ff973b5c78f2"}'
      node.alpha.kubernetes.io/ttl: "0"
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
```

**Comment by brandond** [comment]
> `"--node-ip","10.0.0.1","--advertise-address","10.0.0.1","--node-external-ip","1.2.3.4,10.0.0.1"`

You should not have the same IP as both an internal and an external address. External addresses are meant to be used in situations where a node has a public IP NATed to its private IP address. You might read through the discussion here: https://github.com/k3s-io/k3s/discussions/6029

**Comment by Northcode** [comment]
If I set the external ip to only the public ip, it no longer responds when I try to access it via the internal ip. 
Can I get the LoadBalancer service to respond to both internal and external ip then? 
I would of course like to set external ip to the actual external one like you said of course, but it doesn't seem to be working. 

I'm running an nginx ingress as a LoadBalancer service on port 80 and 443.
If I set `--node-external-ip` to only the external ip, and then try to reach the ingress from node B (10.0.0.2/1.2.3.5) via 10.0.0.1 (the internal address), I get a timeout.

I'd like to do this to be able to filter the ingress based on ip whitelists, so I can have some services, like a docker registry, only accessible from inside the vpn network.
If the LoadBalancer only responds from the public ip, this isn't going to work, since the source ip will be the clients external ip, not the clients vpn ip.

I saw you suggested running the wireguard backend for flannel in the linked discussion, but if I'm not wrong that will still have the same issue? It would also require running wireguard inside of another wireguard, which seems wasteful.

I also saw your comment about changing ips after initial setup was not very supported, could that be causing my issue? I originally had set external-ip to only the vpn ip, 10.0.0.1/2. 
Is there a way to change that properly without having to rebuild the whole cluster? I have a lot of services with state that I would rather like not having to re-setup.

**Comment by brandond** [comment]
> If I set the external ip to only the public ip, it no longer responds when I try to access it via the internal ip.

When you set the external IP to only the external IP, what do you see listed as the external-ip for the LoadBalancer service when you do `kubectl get service -A`?

> I saw you suggested running the wireguard backend for flannel in the linked discussion, but if I'm not wrong that will still have the same issue? It would also require running wireguard inside of another wireguard, which seems wasteful.

No, I was suggesting that you let flannel manage wireguard independently as the Kubernetes network overlay, rather than trying to get flannel to run vxlan over your existing wireguard tunnels.

**Comment by Northcode** [comment]
Ok now this is weird. I was sure I tested that exact configuration on Saturday, and got tons of connection refused errors, but now it seems to be working fine. I have no idea what changed.

So now the external ip of the LB services are just 1.2.3.4/1.2.3.5, but they still respond from inside the vpn network.
So I get the real client ip, just verified that as well.

I tried rebooting them both too to see if there was some non-persistent configuration issue, but its still working, so I have no idea what fixed it.
Thanks for the help.


I've already set nginx to use a DaemonSet, but for other LoadBalancer services I've just pinned them to the public node and set their externaltrafficpolicy to local.
That way I can choose between preserving client ip with HAPROXY-protocol in nginx if it needs to go to the worker node. or, pinning the service to the public-ip master node, in which case the LoadBalancer service preserves ip as long as its external ip is the same as the nodes external ip.


Just following up on your flannel wireguard suggestion though, because I want to understand more about how this all works.
I don't think I know what you mean with letting flannel manage wireguard independently? But not using the flannel backend? Or am I still missunderstanding?
The way I see it there would be 3 ways to make the networking work in my situation (1 public ip + 1 NATed):
- my current solution: wireguard network, flannel as vxlan over that
- wireguard network with flannel as wireguard-backend over that (wireguard in wireguard)
- flannel as wireguard-backend over the internet, sidestepping the other wireguard network altogether, but as far as I'm aware this would require opening both the flannel-wireguard port and the kubernetes management port as public, otherwise how would the worker node connect to the master?
